### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Sh-imaa**
Merging dev branch into master

### CODE REVIEW

File: shared/defaults.yaml [hunk 1]
```diff
@@ -2,13 +2,14 @@
 # -----    Model    -----
 # -----------------------
 model:
-  inchannels: 11
+  inchannels: 10
```
Comment (krisrs1128) [inline shared/defaults.yaml]: Why do you need both this and "channels_to_inc"? Will you ever have the number of in channels not equal to the length of channels_to_inc?

File: src/dataset.py [hunk 2]
```diff
@@ -5,66 +5,101 @@
 import pdb
 
 import torch
-from torch.utils.data import Dataset, DataLoader
+from torch.utils.data import Dataset, Dataloader
```
Comment (krisrs1128) [inline src/dataset.py]: Is this a typo?

File: src/dataset.py [hunk 3]
```diff
@@ -5,66 +5,101 @@
...
+        self.mask_used = mask_used
+
+    def __getitem__(self, i):
+        pathes = ['img_path', 'mask_path', 'border_path']
+        image_path, mask_path, border_path = self.data.iloc[i][pathes]
+
+        image_path = Path(self.base_dir, image_path)
+        mask_path = Path(self.base_dir, mask_path)
+
+        if self.use_cropped:
```
Comment (krisrs1128) [inline src/dataset.py:41]: I wonder if cropping could just a be transform? That seems more direct than saving all the data twice.

File: src/dataset.py [hunk 4]
```diff
@@ -5,66 +5,101 @@
...
+            img = img[self.channels_to_inc]
+
+        if (self.borders) and (not pd.isnull(border_path)):
+            border_path = Path(self.base_dir, border_path)
+            border = np.load(border_path)
+            border = np.expand_dims(border, axis=0)
+            img = np.concatenate((img, border), axis=0)
+            img = torch.from_numpy(img)
+
+
```
Comment (krisrs1128) [inline src/dataset.py]: Can you remove some of the unnecessary whitespace here and in the next few lines?

File: src/dataset.py [hunk 5]
```diff
@@ -5,66 +5,101 @@
...
+        if self.img_transform is not None:
+            img = self.img_transform(img)
+
+        
+
+        # default is 'all glaciers' 
+        # if self.mask_used == 'glaciers': mask = mask
```
Comment (krisrs1128) [inline src/dataset.py]: I think this is left in by mistake.

File: src/exper.py [hunk 6]
```diff
@@ -47,18 +50,22 @@
     wandb.config.update(opts.to_dict())
     wandb.config.update({"__message": parsed_opts.message})
 
+    # load normalization
```
Comment (krisrs1128) [inline src/exper.py]: Is this data on the cluster somewhere? Do you have the script you used to compute it?

File: src/filter_debris_by_perc.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+def get_debris_perc(row, base_folder):
+    mask = np.load(os.path.join(base_folder, row.cropped_label))
+    img = np.load(os.path.join(base_folder, row.cropped_path))
+    img = np.moveaxis(img, -1, 0) # channels first
+    debris_mask = utils.get_debris_glaciers(img, mask)
+    return debris_mask.sum() / mask.sum()
+
+if __name__ == '__main__':
+    parser = ArgumentParser()
+    parser.add_argument("--toy_data", default=True, type=bool,
```
Comment (krisrs1128) [inline src/filter_debris_by_perc.py]: why not make the parser argument just the path to the data? Then you can avoid the if else, and it's also more flexible.

File: src/metrics.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+def recall(pred, true, label=1):
+    tp = ((pred == label) & (true == label)).sum().item()
+    fp = ((pred == label) & (true != label)).sum().item()
+    fn = ((pred != label) & (true == label)).sum().item()
+
+    return tp / (tp + fn)
+
+
+def pixel_acc(pred, true):
+    return (pred == true).sum().item() / true.numel()
```
Comment (krisrs1128) [inline src/metrics.py:31]: is .sum() / .numel() not the same as .mean()?

File: src/exper.py (hunk 6 above)
Comment (krisrs1128) [reply src/exper.py]: Okay, I see the normalization script below. I wonder if you could read the data into a transform method, though, so you could just pass in a list of transforms, rather than having lots of checks.

File: src/trainer.py [hunk 9]
```diff
@@ -21,17 +26,28 @@ def train(self):
...
+    else:
+      loss_f = diceloss()
+
+    metrics = {'pixel_acc': pixel_acc, 'per': precision, 'recall': recall,
+                   'dice': dice, 'iou': IoU}
 
     for epoch in range(self.config.n_epochs):
-      epoch_losses, mean_loss = self.train_epoch(op, loss_f)
-      dev_loss = self.evaluate(loss_f)
+      _, _ = self.train_epoch(op, loss_f)
```
Comment (krisrs1128) [inline src/trainer.py]: if you're not using either of the outputs, why not just

```
self.train_epoch(op, loss_f)
```

File: src/trainer.py [hunk 10]
```diff
@@ -21,17 +26,28 @@ def train(self):
...
+
+      dev_loss, dev_metrics = self.evaluate(loss_f, metrics)
```
Comment (krisrs1128) [inline src/trainer.py:40]: won't these break if `metrics` is none? Because in one case, evaluate returns a tuple, and otherwise it doesn't. Maybe it's safer to just always return a tuple, but if there are no metrics, it just returns an empty dict.

**Review by krisrs1128 — CHANGES_REQUESTED** [review]
Overall this looks good, I only have a few minor typos / nitpicks for you to resolve before I merge it in.

The only larger conceptual part we may want to address later is whether we can make the types of transforms easier to modify / add to, without just creating more arguments and if statements. We could create new transform subclasses, for example. But we don't need to implement that now.

File: shared/defaults.yaml (hunk 1 above)
Comment (Sh-imaa) [reply shared/defaults.yaml]: You can have channels_to_inc = [0, 1] and inchannels = 4 (due to borders and snow index, ..). 
But yes, inchannels can to be calculated from the combination of other options. Will make this change.

File: src/dataset.py (hunk 2 above)
Comment (Sh-imaa) [reply src/dataset.py]: yes. Deleted by mistake and wrote it with a typo! Thanks

File: src/dataset.py (hunk 3 above)
Comment (Sh-imaa) [reply src/dataset.py:41]: We'd have to carry around the vector data of multiple countries and use the tiff for that in the training, it's masking not just cropping. We can save only the cropped version, if we will drop the borders options forever and no need for non-cropped data

File: src/dataset.py (hunk 5 above)
Comment (Sh-imaa) [reply src/dataset.py]: I wrote it in purpose for documentation, but can remove it if it's more confusing.

File: src/metrics.py (hunk 8 above)
Comment (Sh-imaa) [reply src/metrics.py:31]: I'd have to transform from bool to float then take the mean. Either way should work.

File: src/exper.py (hunk 6 above)
Comment (Sh-imaa) [reply src/exper.py]: I'm not sure I got your comment. Do you mean save the transform object? The idea is that we'd have pre-computed mean and std for all channels + slope + snow_index and it can be indexed later based on the the values in the config file

**Comment by Sh-imaa** [comment]
Merging dev branch into master.
Adding support to different classes of glaciers, getting pseudo debris labels and flexible pre-processing scripts.
